Refactor `gitno_extract_url_parts`

ethomson opened the pull request:
`gitno_extract_url_parts` seems a little deficient in a few ways:

* This moves percent decoding into `git_buf`s.  This lets the SSH transport avoid an unnecessary small, temporary `malloc`.
* Now `gitno_extract_url_parts` uses the same buffer-based decoding
* Percent decode usernames even when passwords are not provided
* Percent decode hostnames, per RFC 3986

This builds on #4557.

pks-t commented on src/buffer.c:
```diff
@@ -455,6 +455,37 @@ int git_buf_decode_base85(
+	size_t str_len)
+{
+	size_t str_pos, new_size;
+	int error = 0;
```
This variable is unused

ethomson replied:
Right, it's unnecessary.

pks-t commented on src/buffer.c:
```diff
@@ -455,6 +455,37 @@ int git_buf_decode_base85(
+		if (str[str_pos] == '%' &&
+			str_len > str_pos + 2 &&
+			isxdigit(str[str_pos + 1]) &&
+			isxdigit(str[str_pos + 2])) {
```
I'm still in favour of indenting additional lines for if-statements by four spaces. Makes it more obvious what still belongs to the if-statement and what not. Anyway, no need to settle on that issue now.

pks-t commented on src/buffer.c:
```diff
@@ -455,6 +455,37 @@ int git_buf_decode_base85(
+
+	for (str_pos = 0; str_pos < str_len; buf->size++, str_pos++) {
+		if (str[str_pos] == '%' &&
+			str_len > str_pos + 2 &&
```
Hm. Doesn't this need to be `>=`? I'd expect like this if there is a "%xx" at the end of string it won't be decoded.

ethomson replied:
I don't think so: consider some string of length 4 (`str_len` in this case is `4`).  If we were to use `>=` instead - then this comparison becomes `str_len >= str_pos + 2`, and then this would be true for `str_pos = 2`.  Consider that the string of length 4 is the string: `ab%z`.  At the point in the loop when `str_pos = 2`, then we would evaluate true on the first condition (`str[str_pos] == '%'`), and the second (`str_len >= str_pos + 2`), and then would walk off the end of the array in the fourth.  (`str[str_pos + 2]`).

pks-t replied:
Agreed

pks-t replied:
Convinced, thanks!

pks-t commented on tests/buf/percent.c:
```diff
@@ -0,0 +1,35 @@
+	expect_decode_pass("github.com", "github.com");
+	expect_decode_pass("github.com", "githu%62.com");
+	expect_decode_pass("github.com", "github%2ecom");
+	expect_decode_pass("foo bar baz", "foo%20bar%20baz");
```
You should probably also add the case of trailing "%xx"

ethomson replied:
Sure.

pks-t commented on src/netops.c:
```diff
@@ -206,83 +206,100 @@ void gitno_connection_data_free_ptrs(gitno_connection_data *d)
+		git_buf_oom(&port) ||
+		git_buf_oom(&path) ||
+		git_buf_oom(&username) ||
+		git_buf_oom(&password))
```
This leaks memory in case only some of these buffers OOM'd

ethomson replied:
Yes.

pks-t replied:
Okay

pks-t replied:
Okay

pks-t commented on src/netops.c:
```diff
@@ -206,83 +206,100 @@ void gitno_connection_data_free_ptrs(gitno_connection_data *d)
+	url_userinfo = url + u.field_data[UF_USERINFO].off;
+	url_userinfo_len = u.field_data[UF_USERINFO].len;
+
+	if (has_host)
```
Dunno. Wouldn't this function be easier to read if setting the buffers would be all in one spot? E.g.

```
if (!!(u.field_set & (1 << UF_HOST)))
        git_buf_decode_percent(&host, url + u.field_data[UF_HOST].off, u.field_data[UF_HOST].len);
```

A lot less variables and all handling of that information is at exactly one spot.

ethomson replied:
The lines are soooooo long then.  It is much cleaner to when we use some temporary variables.

pks-t replied:
Certainly true. On the other hand, you could just have locale variables inside of that block

ethomson commented:
I added a test for the trailing `%xx` and changed the tests a bit to ensure that the tests only read the `len` bytes (by adding trailing garbage at the end of the strings, past `len` bytes).  This ensures that we actually look at the `len` bytes instead of looking for a NUL terminator.

pks-t commented on tests/buf/percent.c:
```diff
@@ -0,0 +1,49 @@
+	expect_decode_pass("github.com", "github%2ecom");
+	expect_decode_pass("foo bar baz", "foo%20bar%20baz");
+	expect_decode_pass("foo bar baz", "foo%20bar%20baz");
+	expect_decode_pass("foo bar ", "foo%20bar%20");
```
:+1:

pks-t commented:
Looks good to me. I'm still a bit on the edge with regards to all those variables getting declared at the start of the function. Making them block-local to the conditions would clean that up a bit, I'd guess. Not blocking this PR for it, though. Feel free to merge and do an rc3

ethomson commented:
> Making them block-local to the conditions would clean that up a bit, I'd guess.

Yeah, that seems like a nice compromise.  I did this, and left the `bool` declarations the way they were.  I think it's an improvement.

ethomson commented:
What I don't understand is why the tests succeed _sometimes_.  🤔

pks-t commented:
No idea either. Maybe it depends on the server, where we get redirected to faulty instances by their load balancer in some occasions.

Regarding the compiler I'm completely clueless which one it was. I'm really sure to have experienced it at some point in time, though

ethomson commented:
> Regarding the compiler I'm completely clueless which one it was. I'm really sure to have experienced it at some point in time, though

We have whitespace between decls all over the codebase.  [Here's just one example](https://github.com/libgit2/libgit2/blob/master/tests/merge/trees/commits.c#L29) that has been in the codebase for years now.

If there's a compiler so broken that it can't parse whitespaces, then thankfully nobody's actually used it to compile us or file an issue about it.  We have a lot of problems that keep us from moving things forward that we should be focusing on instead of that.
